### ISSUE TITLE
Unexpected nullable warning in local function by changing declared type of captured local

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#40821]
This is an existing problem, but it is magnified by PR https://github.com/dotnet/roslyn/pull/40755 which infers a nullable type for `var` locals.
I did not manage to repro this in isolation yet. The scenarios I'd come up with were fixed by https://github.com/dotnet/roslyn/pull/40422 already.

Repro:
0. using an enlistment of `master` from `roslyn` repo (which includes the merged PR https://github.com/dotnet/roslyn/pull/40422 for fix to analysis of local functions)
1. change `var` to `Compilation?` in `AbstractTypeImportCompletionService` as shown below
2. `build.cmd -bootstrap` (to use latest compiler bits to compile)
3. get nullable diagnostic

``` diff
diff --git a/src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionService.cs b/src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionService.cs
index 31f60b8..b52778d 100644
--- a/src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionService.cs
+++ b/src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionService.cs
@@ -58,7 +58,7 @@ internal AbstractTypeImportCompletionService(Workspace workspace)
                 return null;
             }

-            var currentCompilation = await currentProject.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+            Compilation? currentCompilation = await currentProject.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
             return getCacheResults.Value.SelectAsArray(GetItemsFromCacheResult);

             ImmutableArray<CompletionItem> GetItemsFromCacheResult(GetCacheResult cacheResult)
```

`D:\repos\roslyn2\src\Features\Core\Portable\Completion\Providers\ImportCompletionProvider\AbstractTypeImportCompletionService.cs(69,26): error CS8602: Dereference of a possibly null reference. [D:\repos\roslyn2\src\Features\Core\Portable\Microsoft.CodeAnalysis.Features.csproj]`

Relates to https://github.com/dotnet/roslyn/pull/40755 which includes a few workarounds in the IDE code to unblock bootstrapping build.

**Comment by RikkiGibson** [comment]
It would be good if we could come up with a minimal reproducer for this bug.

**Comment by jcouv** [comment]
I agree. I tried a couple of things (added [tests](https://github.com/dotnet/roslyn/pull/40755/files#diff-4ef499b331bacc63b10e31e95a00db61)) but they didn't repro.

``` csharp
        [Fact]
        public void VarLocalUsedInLocalFunction()
        {
            var source =
@"#nullable enable
public class C
{
    void M()
    {
        var x = """";
        var y = """";
        System.Action z = local;

        void local() => x.ToString();
        void local2() => y.ToString();
    }
}
";
            var comp = CreateCompilation(source);
            comp.VerifyDiagnostics(
                // (11,14): warning CS8321: The local function 'local2' is declared but never used
                //         void local2() => y.ToString();
                Diagnostic(ErrorCode.WRN_UnreferencedLocalFunction, "local2").WithArguments("local2").WithLocation(11, 14),
                // (11,26): warning CS8602: Dereference of a possibly null reference.
                //         void local2() => y.ToString();
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(11, 26)
                );
        }

        [Fact]
        public void ExplicitNullableLocalUsedInLocalFunction()
        {
            var source =
@"#nullable enable
public class C
{
    void M()
    {
        string? x = """";
        string? y = """";
        System.Action z = local;

        void local() => x.ToString();
        void local2() => y.ToString();
    }
}
";
            var comp = CreateCompilation(source);
            comp.VerifyDiagnostics(
                // (11,14): warning CS8321: The local function 'local2' is declared but never used
                //         void local2() => y.ToString();
                Diagnostic(ErrorCode.WRN_UnreferencedLocalFunction, "local2").WithArguments("local2").WithLocation(11, 14),
                // (11,26): warning CS8602: Dereference of a possibly null reference.
                //         void local2() => y.ToString();
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(11, 26)
                );
        }
```

**Comment by RikkiGibson** [comment]
@jcouv it's not clear to me whether this is a dupe of #40157.

**Comment by jcouv** [comment]
Tried again with original repro steps, but couldn't repro. Closing
